{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "86cc0cdc-ee51-4602-b7fa-8c829cf74bbc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: igraph\n",
      "\n",
      "\n",
      "Attaching package: ‘igraph’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    components\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(Seurat)\n",
    "library(ggplot2)\n",
    "library(CellChat)\n",
    "library(BiocParallel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ed89d3d0-b91b-412d-9630-08d76747f758",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Seurat\n",
      "\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, t\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "adata <-schard::h5ad2seurat('batch_corrected_v1.h5ad', use.raw = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0f302c74-8c87-49b1-ab53-5f664446f97b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata <- NormalizeData(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c8c2fae8-a3ed-41ff-875e-51a5d5d4ec2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Idents(adata) <- \"annot\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "27de5946-39ef-49bc-baa6-d539affe0ad5",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata <- subset(adata, idents = c(\"Acinar Cells\", \"B Cells\", \"Cycling\", \"Normal Duct\"), invert=TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b043c747-ce3e-4eee-9b11-85e9669048db",
   "metadata": {},
   "outputs": [],
   "source": [
    "cond<-SplitObject(adata, split.by = \"condition\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f441489e-7a49-49ca-ac10-e068a82ea9fa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "runCellChat <- function(subset_ad){\n",
    "    assay <- GetAssayData(subset_ad, assay = \"RNA\", slot = \"counts\")\n",
    "    assay <- normalizeData(assay)\n",
    "    Idents(subset_ad) <- \"annot\"\n",
    "    labels <- Idents(subset_ad)\n",
    "    meta <- subset_ad[[]]\n",
    "    meta$samples <- meta$sample_id\n",
    "    meta$labels <- meta$annot\n",
    "    rm(subset_ad)\n",
    "    cellchat <- createCellChat(object = assay, group.by = \"annot\", meta=meta)\n",
    "    CellChatDB <- CellChatDB.mouse\n",
    "    CellChatDB.use <- subsetDB(CellChatDB, search = \"Secreted Signaling\")\n",
    "    cellchat@DB <- CellChatDB.use\n",
    "    cellchat <- subsetData(cellchat)\n",
    "    cellchat <- identifyOverExpressedGenes(cellchat)\n",
    "    cellchat <- identifyOverExpressedInteractions(cellchat)\n",
    "    cellchat <- computeCommunProb(cellchat, type = \"triMean\")\n",
    "    cellchat <- filterCommunication(cellchat, nonFilter.keep=TRUE, rare.keep=TRUE)\n",
    "    cellchat <- computeCommunProbPathway(cellchat)\n",
    "    cellchat <- aggregateNet(cellchat)\n",
    "    cellchat\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e0ddd394-6796-4855-93fe-e39278a89dbc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  Epithelial, Fibroblasts, Granulocytes, Macrophages, NK Cells, T Cells \n",
      "The number of highly variable ligand-receptor pairs used for signaling inference is 976 \n",
      "triMean is used for calculating the average gene expression per cell group. \n",
      "[1] \">>> Run CellChat on sc/snRNA-seq data <<< [2024-10-28 22:16:20.36779]\"\n",
      "  |======================================================================| 100%\n",
      "[1] \">>> CellChat inference is done. Parameter values are stored in `object@options$parameter` <<< [2024-10-28 22:20:20.208735]\"\n",
      "The non-filtered cell-cell communication is stored in `object@net$prob.nonFilter` and `object@net$pval.nonFilter`. \n",
      "\n",
      "\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  Epithelial, Fibroblasts, Granulocytes, Macrophages, NK Cells, T Cells \n",
      "The number of highly variable ligand-receptor pairs used for signaling inference is 1028 \n",
      "triMean is used for calculating the average gene expression per cell group. \n",
      "[1] \">>> Run CellChat on sc/snRNA-seq data <<< [2024-10-28 22:16:23.498958]\"\n",
      "  |======================================================================| 100%\n",
      "[1] \">>> CellChat inference is done. Parameter values are stored in `object@options$parameter` <<< [2024-10-28 22:20:25.32269]\"\n",
      "The non-filtered cell-cell communication is stored in `object@net$prob.nonFilter` and `object@net$pval.nonFilter`. \n",
      "\n",
      "\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  Epithelial, Fibroblasts, Granulocytes, Macrophages, NK Cells, T Cells \n",
      "The number of highly variable ligand-receptor pairs used for signaling inference is 990 \n",
      "triMean is used for calculating the average gene expression per cell group. \n",
      "[1] \">>> Run CellChat on sc/snRNA-seq data <<< [2024-10-28 22:16:26.348606]\"\n",
      "  |======================================================================| 100%\n",
      "[1] \">>> CellChat inference is done. Parameter values are stored in `object@options$parameter` <<< [2024-10-28 22:20:30.775045]\"\n",
      "The non-filtered cell-cell communication is stored in `object@net$prob.nonFilter` and `object@net$pval.nonFilter`. \n",
      "\n",
      "\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "[1] \"Create a CellChat object from a data matrix\"\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "Set cell identities for the new CellChat object \n",
      "The cell groups used for CellChat analysis are  Epithelial, Fibroblasts, Granulocytes, Macrophages, NK Cells, T Cells \n",
      "The number of highly variable ligand-receptor pairs used for signaling inference is 1054 \n",
      "triMean is used for calculating the average gene expression per cell group. \n",
      "[1] \">>> Run CellChat on sc/snRNA-seq data <<< [2024-10-28 22:16:21.283131]\"\n",
      "  |======================================================================| 100%\n",
      "[1] \">>> CellChat inference is done. Parameter values are stored in `object@options$parameter` <<< [2024-10-28 22:20:32.512346]\"\n",
      "The non-filtered cell-cell communication is stored in `object@net$prob.nonFilter` and `object@net$pval.nonFilter`. \n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cellchat_objs <- bplapply(cond, FUN=runCellChat, BPPARAM = MulticoreParam(4))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "01617a21-bb5a-4fca-8a4f-41dda381770b",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(file = \"individ_cellchat.RData\", object = cellchat_objs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "84b18823-1bdb-437a-bc80-5db889e497c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table style=\"width: 100%;\"><tr><td>computeCommunProb {CellChat}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Compute the communication probability/strength between any interacting cell groups</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>To further speed up on large-scale datasets, USER can downsample the data using the function 'subset' from Seurat package (e.g., pbmc.small &lt;- subset(pbmc, downsample = 500)), or using the function 'sketchData' from CellChat, in particular for the large cell clusters;\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre><code class='language-R'>computeCommunProb(\n",
       "  object,\n",
       "  type = c(\"triMean\", \"truncatedMean\", \"thresholdedMean\", \"median\"),\n",
       "  trim = 0.1,\n",
       "  LR.use = NULL,\n",
       "  raw.use = TRUE,\n",
       "  population.size = FALSE,\n",
       "  distance.use = TRUE,\n",
       "  interaction.range = 250,\n",
       "  scale.distance = 0.01,\n",
       "  k.min = 10,\n",
       "  contact.dependent = TRUE,\n",
       "  contact.range = NULL,\n",
       "  contact.knn.k = NULL,\n",
       "  contact.dependent.forced = FALSE,\n",
       "  do.symmetric = TRUE,\n",
       "  nboot = 100,\n",
       "  seed.use = 1L,\n",
       "  Kh = 0.5,\n",
       "  n = 1\n",
       ")\n",
       "</code></pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table>\n",
       "<tr><td><code id=\"object\">object</code></td>\n",
       "<td>\n",
       "<p>CellChat object</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"type\">type</code></td>\n",
       "<td>\n",
       "<p>Methods for computing the average gene expression per cell group. By default = &quot;triMean&quot;, producing fewer but stronger interactions;\n",
       "When setting &lsquo;type = &quot;truncatedMean&quot;', a value should be assigned to &rsquo;trim',  producing more interactions.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"trim\">trim</code></td>\n",
       "<td>\n",
       "<p>the fraction (0 to 0.25) of observations to be trimmed from each end of x before the mean is computed</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"LR.use\">LR.use</code></td>\n",
       "<td>\n",
       "<p>A subset of ligand-receptor interactions used in inferring communication network</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"raw.use\">raw.use</code></td>\n",
       "<td>\n",
       "<p>Whether use the raw data (i.e., 'object@data.signaling') or the smoothed data (i.e., 'object@data.smooth').\n",
       "Set raw.use = FALSE to use the projected data when analyzing single-cell data with shallow sequencing depth because the projected data could help to reduce the dropout effects of signaling genes, in particular for possible zero expression of subunits of ligands/receptors.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"population.size\">population.size</code></td>\n",
       "<td>\n",
       "<p>Whether consider the proportion of cells in each group across all sequenced cells.\n",
       "Set population.size = FALSE if analyzing sorting-enriched single cells, to remove the potential artifact of population size.\n",
       "Set population.size = TRUE if analyzing unsorted single-cell transcriptomes, with the reason that abundant cell populations tend to send collectively stronger signals than the rare cell populations.\n",
       "</p>\n",
       "<p>Parameters for spatial data analysis:</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"distance.use\">distance.use</code></td>\n",
       "<td>\n",
       "<p>Whether to use distance constraints to compute communication probability. Setting 'distance.use = TRUE' indicates that the cell-cell communication probability is inversely proportional to the computed distance.\n",
       "Setting 'distance.use = FALSE' will only filter out interactions between spatially distant regions, but not add distance constraints.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"interaction.range\">interaction.range</code></td>\n",
       "<td>\n",
       "<p>The maximum interaction/diffusion length of ligands (Unit: microns). This hard threshold is used to filter out the connections between spatially distant regions</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"scale.distance\">scale.distance</code></td>\n",
       "<td>\n",
       "<p>A scale or normalization factor for the spatial distances when setting 'distance.use = TRUE'. For example, scale.distance equals 1, 0.1, 0.01, 0.001, 0.11, or 0.011. We choose this values such that the minimum value of the scaled distances is in [1,2]. This value is not necessary when setting 'distance.use = FALSE'.\n",
       "</p>\n",
       "<p>When comparing communication across different CellChat objects, the same scale factor should be used. For a single CellChat analysis, different scale factors will not affect the ranking of the signaling based on their interaction strength.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"k.min\">k.min</code></td>\n",
       "<td>\n",
       "<p>The minimum number of interacting cell pairs required for defining spatially proximal cell groups.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"contact.dependent\">contact.dependent</code></td>\n",
       "<td>\n",
       "<p>Whether using the 'contact-dependent' manner for inference signaling, that is determining interacting cell pairs by requiring cells to be in direct membrane-membrane contact. By default 'contact.dependent = TRUE' when inferring contact-dependent and juxtacrine signaling (that is &quot;Cell-Cell Contact&quot; signaling classified in CellChatDB$interaction$annotation).\n",
       "If only focusing on 'Secreted Signaling', the 'contact-dependent' manner will be not used except for setting 'contact.dependent.forced = TRUE'.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"contact.range\">contact.range</code></td>\n",
       "<td>\n",
       "<p>The interaction range (Unit: microns) to restrict the contact-dependent signaling when 'contact.dependent = TRUE'.\n",
       "For spatial transcriptomics in a single-cell resolution, 'contact.range' is approximately equal to the estimated cell diameter (i.e., the cell center-to-center distance), which means that contact-dependent and juxtacrine signaling can only happens when the two cells are contact to each other.\n",
       "</p>\n",
       "<p>Typically, 'contact.range = 10', which is a typical human cell size. However, for low-resolution spatial data such as 10X visium, it should be the cell center-to-center distance (i.e., 'contact.range = 100' for visium data).  The function 'computeCellDistance' can compute the center-to-center distance.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"contact.knn.k\">contact.knn.k</code></td>\n",
       "<td>\n",
       "<p>Number of neighbors to restrict the contact-dependent signaling within the neatest neighbors when 'contact.dependent = TRUE'. By default, CellChat uses 'contact.range' to restrict the contact-dependent signaling; however, users can also provide a value of 'contact.knn.k', in order to determine interacting cell pairs based on the k-nearest neighbors (knn).\n",
       "For 10X visium, contact.knn.k = 6. For other spatial technologies, this value may be hard to determine because the sequenced cells/spots are usually not regularly arranged.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"contact.dependent.forced\">contact.dependent.forced</code></td>\n",
       "<td>\n",
       "<p>Whether forcing to use the 'contact-dependent' manner for inference signaling for all L-R pairs including secreted signaling. Users can set 'contact.dependent.forced = TRUE' if also preferring interactions within a contact manner for 'Secreted Signaling'.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"do.symmetric\">do.symmetric</code></td>\n",
       "<td>\n",
       "<p>Whether converting the adjacent matrix into symmetric one when determining spatially proximal cell groups. Default is TRUE, indicating that if adj(i,j) or adj(j,i) is zero, then both are zeros.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"nboot\">nboot</code></td>\n",
       "<td>\n",
       "<p>Threshold of p-values</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"seed.use\">seed.use</code></td>\n",
       "<td>\n",
       "<p>Set a random seed. By default, set the seed to 1.</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"Kh\">Kh</code></td>\n",
       "<td>\n",
       "<p>Parameter in Hill function</p>\n",
       "</td></tr>\n",
       "<tr><td><code id=\"n\">n</code></td>\n",
       "<td>\n",
       "<p>Parameter in Hill function</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>A CellChat object with updated slot 'net':\n",
       "</p>\n",
       "<p>object@net$prob is the inferred communication probability (strength) array, where the first, second and third dimensions represent a source, target and ligand-receptor pair, respectively.\n",
       "</p>\n",
       "<p>USER can access all the inferred cell-cell communications using the function 'subsetCommunication(object)', which returns a data frame.\n",
       "</p>\n",
       "<p>object@net$pval is the corresponding p-values of each interaction\n",
       "</p>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>CellChat</em> version 2.1.2 ]</div></main>\n",
       "\n",
       "</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{computeCommunProb}{Compute the communication probability/strength between any interacting cell groups}{computeCommunProb}\n",
       "%\n",
       "\\begin{Description}\n",
       "To further speed up on large-scale datasets, USER can downsample the data using the function 'subset' from Seurat package (e.g., pbmc.small <- subset(pbmc, downsample = 500)), or using the function `sketchData` from CellChat, in particular for the large cell clusters;\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "computeCommunProb(\n",
       "  object,\n",
       "  type = c(\"triMean\", \"truncatedMean\", \"thresholdedMean\", \"median\"),\n",
       "  trim = 0.1,\n",
       "  LR.use = NULL,\n",
       "  raw.use = TRUE,\n",
       "  population.size = FALSE,\n",
       "  distance.use = TRUE,\n",
       "  interaction.range = 250,\n",
       "  scale.distance = 0.01,\n",
       "  k.min = 10,\n",
       "  contact.dependent = TRUE,\n",
       "  contact.range = NULL,\n",
       "  contact.knn.k = NULL,\n",
       "  contact.dependent.forced = FALSE,\n",
       "  do.symmetric = TRUE,\n",
       "  nboot = 100,\n",
       "  seed.use = 1L,\n",
       "  Kh = 0.5,\n",
       "  n = 1\n",
       ")\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{object}] CellChat object\n",
       "\n",
       "\\item[\\code{type}] Methods for computing the average gene expression per cell group. By default = \"triMean\", producing fewer but stronger interactions;\n",
       "When setting `type = \"truncatedMean\"`, a value should be assigned to 'trim',  producing more interactions.\n",
       "\n",
       "\\item[\\code{trim}] the fraction (0 to 0.25) of observations to be trimmed from each end of x before the mean is computed\n",
       "\n",
       "\\item[\\code{LR.use}] A subset of ligand-receptor interactions used in inferring communication network\n",
       "\n",
       "\\item[\\code{raw.use}] Whether use the raw data (i.e., `object@data.signaling`) or the smoothed data (i.e., `object@data.smooth`).\n",
       "Set raw.use = FALSE to use the projected data when analyzing single-cell data with shallow sequencing depth because the projected data could help to reduce the dropout effects of signaling genes, in particular for possible zero expression of subunits of ligands/receptors.\n",
       "\n",
       "\\item[\\code{population.size}] Whether consider the proportion of cells in each group across all sequenced cells.\n",
       "Set population.size = FALSE if analyzing sorting-enriched single cells, to remove the potential artifact of population size.\n",
       "Set population.size = TRUE if analyzing unsorted single-cell transcriptomes, with the reason that abundant cell populations tend to send collectively stronger signals than the rare cell populations.\n",
       "\n",
       "Parameters for spatial data analysis:\n",
       "\n",
       "\\item[\\code{distance.use}] Whether to use distance constraints to compute communication probability. Setting `distance.use = TRUE` indicates that the cell-cell communication probability is inversely proportional to the computed distance.\n",
       "Setting `distance.use = FALSE` will only filter out interactions between spatially distant regions, but not add distance constraints.\n",
       "\n",
       "\\item[\\code{interaction.range}] The maximum interaction/diffusion length of ligands (Unit: microns). This hard threshold is used to filter out the connections between spatially distant regions\n",
       "\n",
       "\\item[\\code{scale.distance}] A scale or normalization factor for the spatial distances when setting `distance.use = TRUE`. For example, scale.distance equals 1, 0.1, 0.01, 0.001, 0.11, or 0.011. We choose this values such that the minimum value of the scaled distances is in [1,2]. This value is not necessary when setting `distance.use = FALSE`.\n",
       "\n",
       "When comparing communication across different CellChat objects, the same scale factor should be used. For a single CellChat analysis, different scale factors will not affect the ranking of the signaling based on their interaction strength.\n",
       "\n",
       "\\item[\\code{k.min}] The minimum number of interacting cell pairs required for defining spatially proximal cell groups.\n",
       "\n",
       "\\item[\\code{contact.dependent}] Whether using the `contact-dependent` manner for inference signaling, that is determining interacting cell pairs by requiring cells to be in direct membrane-membrane contact. By default `contact.dependent = TRUE` when inferring contact-dependent and juxtacrine signaling (that is \"Cell-Cell Contact\" signaling classified in CellChatDB\\$interaction\\$annotation).\n",
       "If only focusing on `Secreted Signaling`, the `contact-dependent` manner will be not used except for setting `contact.dependent.forced = TRUE`.\n",
       "\n",
       "\\item[\\code{contact.range}] The interaction range (Unit: microns) to restrict the contact-dependent signaling when `contact.dependent = TRUE`.\n",
       "For spatial transcriptomics in a single-cell resolution, `contact.range` is approximately equal to the estimated cell diameter (i.e., the cell center-to-center distance), which means that contact-dependent and juxtacrine signaling can only happens when the two cells are contact to each other.\n",
       "\n",
       "Typically, `contact.range = 10`, which is a typical human cell size. However, for low-resolution spatial data such as 10X visium, it should be the cell center-to-center distance (i.e., `contact.range = 100` for visium data).  The function `computeCellDistance` can compute the center-to-center distance.\n",
       "\n",
       "\\item[\\code{contact.knn.k}] Number of neighbors to restrict the contact-dependent signaling within the neatest neighbors when `contact.dependent = TRUE`. By default, CellChat uses `contact.range` to restrict the contact-dependent signaling; however, users can also provide a value of `contact.knn.k`, in order to determine interacting cell pairs based on the k-nearest neighbors (knn).\n",
       "For 10X visium, contact.knn.k = 6. For other spatial technologies, this value may be hard to determine because the sequenced cells/spots are usually not regularly arranged.\n",
       "\n",
       "\\item[\\code{contact.dependent.forced}] Whether forcing to use the `contact-dependent` manner for inference signaling for all L-R pairs including secreted signaling. Users can set `contact.dependent.forced = TRUE` if also preferring interactions within a contact manner for `Secreted Signaling`.\n",
       "\n",
       "\\item[\\code{do.symmetric}] Whether converting the adjacent matrix into symmetric one when determining spatially proximal cell groups. Default is TRUE, indicating that if adj(i,j) or adj(j,i) is zero, then both are zeros.\n",
       "\n",
       "\\item[\\code{nboot}] Threshold of p-values\n",
       "\n",
       "\\item[\\code{seed.use}] Set a random seed. By default, set the seed to 1.\n",
       "\n",
       "\\item[\\code{Kh}] Parameter in Hill function\n",
       "\n",
       "\\item[\\code{n}] Parameter in Hill function\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Value}\n",
       "A CellChat object with updated slot 'net':\n",
       "\n",
       "object@net\\$prob is the inferred communication probability (strength) array, where the first, second and third dimensions represent a source, target and ligand-receptor pair, respectively.\n",
       "\n",
       "USER can access all the inferred cell-cell communications using the function 'subsetCommunication(object)', which returns a data frame.\n",
       "\n",
       "object@net\\$pval is the corresponding p-values of each interaction\n",
       "\\end{Value}"
      ],
      "text/plain": [
       "computeCommunProb           package:CellChat           R Documentation\n",
       "\n",
       "_\bC_\bo_\bm_\bp_\bu_\bt_\be _\bt_\bh_\be _\bc_\bo_\bm_\bm_\bu_\bn_\bi_\bc_\ba_\bt_\bi_\bo_\bn _\bp_\br_\bo_\bb_\ba_\bb_\bi_\bl_\bi_\bt_\by/_\bs_\bt_\br_\be_\bn_\bg_\bt_\bh _\bb_\be_\bt_\bw_\be_\be_\bn _\ba_\bn_\by _\bi_\bn_\bt_\be_\br_\ba_\bc_\bt_\bi_\bn_\bg\n",
       "_\bc_\be_\bl_\bl _\bg_\br_\bo_\bu_\bp_\bs\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     To further speed up on large-scale datasets, USER can downsample\n",
       "     the data using the function 'subset' from Seurat package (e.g.,\n",
       "     pbmc.small <- subset(pbmc, downsample = 500)), or using the\n",
       "     function `sketchData` from CellChat, in particular for the large\n",
       "     cell clusters;\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     computeCommunProb(\n",
       "       object,\n",
       "       type = c(\"triMean\", \"truncatedMean\", \"thresholdedMean\", \"median\"),\n",
       "       trim = 0.1,\n",
       "       LR.use = NULL,\n",
       "       raw.use = TRUE,\n",
       "       population.size = FALSE,\n",
       "       distance.use = TRUE,\n",
       "       interaction.range = 250,\n",
       "       scale.distance = 0.01,\n",
       "       k.min = 10,\n",
       "       contact.dependent = TRUE,\n",
       "       contact.range = NULL,\n",
       "       contact.knn.k = NULL,\n",
       "       contact.dependent.forced = FALSE,\n",
       "       do.symmetric = TRUE,\n",
       "       nboot = 100,\n",
       "       seed.use = 1L,\n",
       "       Kh = 0.5,\n",
       "       n = 1\n",
       "     )\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "  object: CellChat object\n",
       "\n",
       "    type: Methods for computing the average gene expression per cell\n",
       "          group. By default = \"triMean\", producing fewer but stronger\n",
       "          interactions; When setting `type = \"truncatedMean\"`, a value\n",
       "          should be assigned to 'trim', producing more interactions.\n",
       "\n",
       "    trim: the fraction (0 to 0.25) of observations to be trimmed from\n",
       "          each end of x before the mean is computed\n",
       "\n",
       "  LR.use: A subset of ligand-receptor interactions used in inferring\n",
       "          communication network\n",
       "\n",
       " raw.use: Whether use the raw data (i.e., `object@data.signaling`) or\n",
       "          the smoothed data (i.e., `object@data.smooth`). Set raw.use =\n",
       "          FALSE to use the projected data when analyzing single-cell\n",
       "          data with shallow sequencing depth because the projected data\n",
       "          could help to reduce the dropout effects of signaling genes,\n",
       "          in particular for possible zero expression of subunits of\n",
       "          ligands/receptors.\n",
       "\n",
       "population.size: Whether consider the proportion of cells in each group\n",
       "          across all sequenced cells. Set population.size = FALSE if\n",
       "          analyzing sorting-enriched single cells, to remove the\n",
       "          potential artifact of population size. Set population.size =\n",
       "          TRUE if analyzing unsorted single-cell transcriptomes, with\n",
       "          the reason that abundant cell populations tend to send\n",
       "          collectively stronger signals than the rare cell populations.\n",
       "\n",
       "          Parameters for spatial data analysis:\n",
       "\n",
       "distance.use: Whether to use distance constraints to compute\n",
       "          communication probability. Setting `distance.use = TRUE`\n",
       "          indicates that the cell-cell communication probability is\n",
       "          inversely proportional to the computed distance. Setting\n",
       "          `distance.use = FALSE` will only filter out interactions\n",
       "          between spatially distant regions, but not add distance\n",
       "          constraints.\n",
       "\n",
       "interaction.range: The maximum interaction/diffusion length of ligands\n",
       "          (Unit: microns). This hard threshold is used to filter out\n",
       "          the connections between spatially distant regions\n",
       "\n",
       "scale.distance: A scale or normalization factor for the spatial\n",
       "          distances when setting `distance.use = TRUE`. For example,\n",
       "          scale.distance equals 1, 0.1, 0.01, 0.001, 0.11, or 0.011. We\n",
       "          choose this values such that the minimum value of the scaled\n",
       "          distances is in [1,2]. This value is not necessary when\n",
       "          setting `distance.use = FALSE`.\n",
       "\n",
       "          When comparing communication across different CellChat\n",
       "          objects, the same scale factor should be used. For a single\n",
       "          CellChat analysis, different scale factors will not affect\n",
       "          the ranking of the signaling based on their interaction\n",
       "          strength.\n",
       "\n",
       "   k.min: The minimum number of interacting cell pairs required for\n",
       "          defining spatially proximal cell groups.\n",
       "\n",
       "contact.dependent: Whether using the `contact-dependent` manner for\n",
       "          inference signaling, that is determining interacting cell\n",
       "          pairs by requiring cells to be in direct membrane-membrane\n",
       "          contact. By default `contact.dependent = TRUE` when inferring\n",
       "          contact-dependent and juxtacrine signaling (that is\n",
       "          \"Cell-Cell Contact\" signaling classified in\n",
       "          CellChatDB$interaction$annotation). If only focusing on\n",
       "          `Secreted Signaling`, the `contact-dependent` manner will be\n",
       "          not used except for setting `contact.dependent.forced =\n",
       "          TRUE`.\n",
       "\n",
       "contact.range: The interaction range (Unit: microns) to restrict the\n",
       "          contact-dependent signaling when `contact.dependent = TRUE`.\n",
       "          For spatial transcriptomics in a single-cell resolution,\n",
       "          `contact.range` is approximately equal to the estimated cell\n",
       "          diameter (i.e., the cell center-to-center distance), which\n",
       "          means that contact-dependent and juxtacrine signaling can\n",
       "          only happens when the two cells are contact to each other.\n",
       "\n",
       "          Typically, `contact.range = 10`, which is a typical human\n",
       "          cell size. However, for low-resolution spatial data such as\n",
       "          10X visium, it should be the cell center-to-center distance\n",
       "          (i.e., `contact.range = 100` for visium data).  The function\n",
       "          `computeCellDistance` can compute the center-to-center\n",
       "          distance.\n",
       "\n",
       "contact.knn.k: Number of neighbors to restrict the contact-dependent\n",
       "          signaling within the neatest neighbors when\n",
       "          `contact.dependent = TRUE`. By default, CellChat uses\n",
       "          `contact.range` to restrict the contact-dependent signaling;\n",
       "          however, users can also provide a value of `contact.knn.k`,\n",
       "          in order to determine interacting cell pairs based on the\n",
       "          k-nearest neighbors (knn). For 10X visium, contact.knn.k = 6.\n",
       "          For other spatial technologies, this value may be hard to\n",
       "          determine because the sequenced cells/spots are usually not\n",
       "          regularly arranged.\n",
       "\n",
       "contact.dependent.forced: Whether forcing to use the\n",
       "          `contact-dependent` manner for inference signaling for all\n",
       "          L-R pairs including secreted signaling. Users can set\n",
       "          `contact.dependent.forced = TRUE` if also preferring\n",
       "          interactions within a contact manner for `Secreted\n",
       "          Signaling`.\n",
       "\n",
       "do.symmetric: Whether converting the adjacent matrix into symmetric one\n",
       "          when determining spatially proximal cell groups. Default is\n",
       "          TRUE, indicating that if adj(i,j) or adj(j,i) is zero, then\n",
       "          both are zeros.\n",
       "\n",
       "   nboot: Threshold of p-values\n",
       "\n",
       "seed.use: Set a random seed. By default, set the seed to 1.\n",
       "\n",
       "      Kh: Parameter in Hill function\n",
       "\n",
       "       n: Parameter in Hill function\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     A CellChat object with updated slot 'net':\n",
       "\n",
       "     object@net$prob is the inferred communication probability\n",
       "     (strength) array, where the first, second and third dimensions\n",
       "     represent a source, target and ligand-receptor pair, respectively.\n",
       "\n",
       "     USER can access all the inferred cell-cell communications using\n",
       "     the function 'subsetCommunication(object)', which returns a data\n",
       "     frame.\n",
       "\n",
       "     object@net$pval is the corresponding p-values of each interaction\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?computeCommunProb"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
